vlad_task3/DynamicArrayProxyTest.cpp

```diff
@@ -0,0 +1,125 @@
+
+#include <boost/test/unit_test.hpp>
+#include <iostream>
+#include <Common.hpp>
+#include <ServerProxyObject.hpp>
+#include <ClientProxyObject.hpp>
+#include <unistd.h>
+#include <thread>
+#include <sys/stat.h>
+#include <signal.h>
+#include <fstream>
+#include <boost/format.hpp>
+#include <cstring>
+
+using namespace common;
+
+struct Fixture
+{
+    const char *const client_input_pipes[3] = {"/tmp/cat_client_input_0", "/tmp/cat_client_input_1"};
+    Fixture()
+    {
+        remove(client_input_pipes[0]);
+        remove(client_input_pipes[1]);
+        mkfifo(client_input_pipes[0], S_IWUSR | S_IRUSR | S_IROTH);
+        mkfifo(client_input_pipes[1], S_IWUSR | S_IRUSR | S_IROTH);
+    }
+    ~Fixture()
+    {
+        remove(client_input_pipes[0]);
+        remove(client_input_pipes[1]);
+    }
+    int count = 100;
+    char *trim(char *s)
+    {
+        char *back = s + strlen(s);
+        while (isspace(*--back))
+            ;
+        *(back + 1) = '\0';
+        return s;
+    }
+};
+
+BOOST_FIXTURE_TEST_CASE(DynamicArrayProxy_single_client, Fixture)
+{
+    std::unique_ptr<FILE, decltype(&pclose)> serv(popen("./main_server", "r"), pclose);
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    std::string cmd = (boost::format("./main_client --no-prompt < %s") % client_input_pipes[0]).str();
+    std::unique_ptr<FILE, decltype(&pclose)> client(popen(cmd.c_str(), "r"), pclose);
+    std::ofstream input_pipe(client_input_pipes[0]);
+    input_pipe << "add\nstr1\nget\n0\ndelete\nstr1\nexit" << std::endl;
+    char line[100] = {0};
+
+    fgets(line, 100, client.get());
+    BOOST_CHECK(!strcmp(line, "String added successfully\n"));
+    memset(line, 0, sizeof(line));
+    fgets(line, 100, client.get());
+    BOOST_CHECK(!strcmp(line, "arr[0]=\"str1\"\n"));
+    memset(line, 0, sizeof(line));
+    fgets(line, 100, client.get());
+    BOOST_CHECK(!strcmp(line, "String deleted successfully\n"));
+}
+
+BOOST_FIXTURE_TEST_CASE(DynamicArrayProxy_multiple_clients, Fixture)
+{
+    std::unique_ptr<FILE, decltype(&pclose)> serv(popen("./main_server", "r"), pclose);
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    std::string cmd1 = (boost::format("./main_client --no-prompt < %s") % client_input_pipes[0]).str();
+    std::string cmd2 = (boost::format("./main_client --no-prompt < %s") % client_input_pipes[1]).str();
+    std::unique_ptr<FILE, decltype(&pclose)> client1(popen(cmd1.c_str(), "r"), pclose);
+    std::unique_ptr<FILE, decltype(&pclose)> client2(popen(cmd2.c_str(), "r"), pclose);
+
+    std::ofstream input1(client_input_pipes[0]);
+    std::ofstream input2(client_input_pipes[1]);
+
+    for (int i = 0; i < 100; i++)
+    {
+        std::string str1 = (boost::format("add\nstr1%d\n") % i).str();
+        input1 << str1 << std::endl;
+        std::string str2 = (boost::format("add\nstr2%d\n") % i).str();
+        input2 << str2 << std::endl;
+    }
+    for (int i = 0; i < 100; i++)
```

Comment:
This is not a parallel execution, maybe if you replace the for with for_each will work.